### ISSUE TITLE
Custom Data Input

### ISSUE THREAD

**Issue by mgarbade** [Russell91/LSTMSummation#1]
I was going through your LSTM example (It was quite hard to understand for me since the layers are executed in a loop over the function lstm_layers... but I'm also new to LSTM and numpy). 
Now to my question:

1) How can I use custom (hdf5) data instead of randomly generated numpy data? I know how to do it in Caffe but I didn't know where to look it up for Apollocaffe. Can you please just provide a quick example for reading in Hdf5 data?

2) Is there some documentation where I can look up these things?

**Comment by Russell91** [comment]
Well, you would do it the same way in ApolloCaffe that you do it in Caffe. Either pass net.f('''your hdf5 prototxt layer string'''), or you could also use apollocaffe.layers.HDF5DataLayer. It's up to you really. Unfortunately, we don't have documentation covering every part of Caffe, but all the apollocaffe.layers classes have almost exactly the same API, so even playing around with the HDF5DataLayer class for 5 minutes should make it clear how to use it.

**Comment by mgarbade** [comment]
Sorry but I could not find a corresponding documentation in the caffe examples. So the file "caffe_layers.py" which includes the "HDF5Data" class looks like this:

    class HDF5Data(DataLayer):
        def __init__(self, name, source, batch_size, transform=None, **kwargs):

Which doesn't tell me much...

When using caffe for training I just modify the data-layer like this:

    name: "mnist"
    type: "HDF5Data"
    top: "data"
    top: "label"
    include {
      phase: TRAIN
    }
    hdf5_data_param {
      source: "/path/to/database_file.txt"
      batch_size: 250
    }

So I tried to adapt your toy example. Can you just tell me what arguments I should give the HDF5Data function? (Just putting  "/path/to/database_file.txt" there doesn't seem to work)

    net = apollocaffe.ApolloNet()
    for i in range(1000):
        example = np.array(np.random.random()).reshape((1, 1, 1, 1))
        net.clear_forward()
        net.f(HDF5Data('data', ???))
        net.f(HDF5Data('label', ???))
        net.f(Convolution('conv', (1,1), 1, bottoms=['data']))
        net.f(EuclideanLoss('loss', bottoms=['conv', 'label']))
        net.backward()
        net.update(lr=0.1)
        if i % 100 == 0:
            print net.loss

Sorry for asking you basically implementation questions but I'm not sure where else to get help. I would  really like to look these things up somewhere else. But to me the caffe python notebook examples look quite differently. Also doing things like "HDFData?" doesn't give me any example or documentation and as explained above: Looking at the *.py source code seems not to be self explanatory either.

**Comment by Russell91** [comment]
layer = HDF5Data('mnist', '/path/to/database_file.txt', 250, tops=['data', 'label'])

This returns a google protobuf object, which you can't print with:

    print layer.p

to verify that it is equal to the string you want. You'll need to adapt the include: TRAIN part yourself by using the correct HDF5Data layer based on when you're training vs testing.

**Comment by mgarbade** [comment]
Thank you so much. I finally got a rough version that trains and tests on my HDF5 data to work :) :

	import apollocaffe
	from apollocaffe.layers import HDF5Data, InnerProduct, ReLU, SoftmaxWithLoss, Accuracy

	train_data = HDF5Data('mnist', '/path/to/train_database_file.txt', 250, tops=['data', 'label'])
	test_data = HDF5Data('mnist', '/path/to/test_database_file.txt', 800, tops=['data', 'label'])
	accuracy = Accuracy('acc', bottoms=['ip2', 'label'])
	ip1 = InnerProduct('ip1',4000,bottoms=["data"])
	relu1 = ReLU('relu1',tops=["ip1"],bottoms=["ip1"])
	ip2 = InnerProduct('ip2',25,bottoms=["ip1"])
	loss = SoftmaxWithLoss('loss', bottoms=['ip2', 'label'])

	# Train Network
	net = apollocaffe.ApolloNet()
	for i in range(10000):
	    net.clear_forward()
	    net.f(train_data)
	    net.f(ip1)
	    net.f(relu1)
	    net.f(ip2)
	    net.f(loss)
	    net.backward()
	    net.update(lr=0.01)
	    if i % 1000 == 0:
		print net.loss


	# Test Accuracy on Validation Set
	predictions = []
	for i in range(10):
	    net.clear_forward()
	    net.f(testDataLayer)
	    net.f(iplayer1)
	    net.f(relu1)
	    net.f(iplayer2)
	    net.f(accuracyLayer)
	    predictions.append(float(net.blobs["acc"].data.flatten()[0]))


	print np.average(predictions)


My next goal is to run a sequence of input data through an additional LSTM layer. You provided an "LSTM-Tutorial" for that but again its hard for me to extract the gist from sample. I would basically need to throw away the "numpy-random-data-generation-part" and train it with data stored in a HDF5 database.

1)  Could you give a hint of how to adapt the "LSTM-Tutorial" to custom (HDF5) input data?

In "LSTM-Tutorial" inside the function `forward(net,net_config)`, there are two loops. The first `for step in range(length):` goes over every number inside the input sequence. The second executes the forward pass of the `LstmUnit`
    for l in lstm_layers(net, step, filler, net_config):
                net.f(l) 

2) How often is this loop called ? Is it always called twice? (Once for the `Concat` unit and another time for the `LstmUnit`?

**Comment by mgarbade** [comment]
So I'm moving forward slowly :)

to 2) yes 

to 1)  Here is a small example for training a Lstm Neural network using custom hdf5 input data:

First to generate random training data and store them in the hdf5 format using Matlab:

    % Create a sequence of length 5 of integer numbers in the range [1:10] for training and testing
    train_data = randi(10,1000,5);
    test_data = randi(10,10,5);

    function [labels] = createLstmTutDataset(data)
      numSamples = size(data,1);
      labels = zeros(numSamples,1);

      for i = 1:numSamples
	  % labels :: category: [0  -> more small vals (<5)] [1 -> more big vals (>=5)]
	  if (sum(featVec >= 5) >= 3 )
	      labels(i) = 1;
	  else
	      labels(i) = 0;
	  end
	  
      end
    end

Now to train the Lstm network in python:

    import h5py
    import json
    import random
    import numpy as np
    import apollocaffe
    from apollocaffe.layers import (HDF5Data, InnerProduct, ReLU, SoftmaxWithLoss, Accuracy, NumpyData,
				    LstmUnit, Concat, EuclideanLoss, Filler)

    

    config = json.load(open("config.json", 'r'))
    apollocaffe.set_random_seed(config["solver"]["random_seed"])
    
    # Load training data from hdf5 database 
    h5f = h5py.File('train.h5','r')
    train_data = h5f['/data'][:]
    train_label = h5f['/label'][:]
    
    # Define training parameters
    net_config = config["net"]
    numSamplesTrain = train_data.shape[0]
    batch_size = 10
    sequence_length = 5
    numTrainingIterations = 30000

    net = apollocaffe.ApolloNet()
    for i in range(numTrainingIterations):
	net.clear_forward()
	# Get a random batch from the training data
	rowIdx = random.sample(range(1, numSamplesTrain), batch_size)
	batch_labels = train_label[rowIdx]
	batch_data = train_data[rowIdx,:]
	
	# Start to process the training batch
	filler = Filler("uniform", net_config["init_range"])
	net.f(NumpyData("lstm_seed", np.zeros((batch_size, net_config["mem_cells"]))))
	for step in range(sequence_length):
	    value = batch_data[:,step]
	    net.f(NumpyData("value%d" % step,value.reshape((-1, 1)))) 
	    if step == 0:
		prev_hidden = "lstm_seed"
		prev_mem = "lstm_seed"
	    else:
		prev_hidden = "lstm_hidden%d" % (step - 1)
		prev_mem = "lstm_mem%d" % (step - 1)
	    net.f(Concat("lstm_concat%d" % step,bottoms=[prev_hidden, "value%d" % step]))
	    net.f(LstmUnit("lstm%d" % step, net_config["mem_cells"],
		bottoms=["lstm_concat%d" % step, prev_mem],
		param_names=["input_value", "input_gate", "forget_gate", "output_gate"],
		tops=["lstm_hidden%d" % step, "lstm_mem%d" % step],
		weight_filler=filler))

	net.f(InnerProduct("ip", 2, bottoms=["lstm_hidden%d" % (sequence_length - 1)],weight_filler=filler))
	net.f(NumpyData("label", np.reshape(batch_labels , (-1, 1))))
	net.f(SoftmaxWithLoss('loss', bottoms=['ip', 'label']))
	net.backward()
	net.update(lr=0.01)
	if i % 1000 == 0:
	    print net.loss



The problem is that the data is still 1-dimensional. I tried to adapt the the example to higher dimensions but failed so far.



Can you give me a hint how to modify this example such that it works with input data of higher dimensions?
At the moment I'm getting an error from the Concat layer, saying that I'm either trying to concatenate the wrong axes or the number of axes of the blobs to be concatenated to not match.
i) Do I have to modify the shape of the hidden and memory layers? 
  So far the dimensions of "lstm_seed" are [  batch_size, net_config["mem_cells"]  ]. What would be the dimesions if the input of every frame of a sequences was say 14?
  a) [ batch_size, net_config["mem_cells"], 14  ]
  or 
  b) [ batch_size, 14, net_config["mem_cells"]  ]

ii) Is the LstmUnit capable of processing input data of higher dimension?

Sorry again for asking these implementation questions. I would really like to use your module efficiently, so any small tip would help me a lot.

**Comment by Russell91** [comment]
Looks like you've been making good progress! Please be aware that Apollocaffe provides a simple Reshape layer written in python. You can use it like this:

    net.f(Reshape('bottom_reshaped', (batch_size, 14, 1), bottoms=['bottom_of_shape_batch_size_x_1']))

Hopefully that helps.
